test/distributed/fsdp/test_fsdp_backward_prefetch.py

```diff
@@ -0,0 +1,202 @@
+    "decoder.layers.{index}.multihead_attn.out_proj.bias",
+    "decoder.layers.{index}.linear1.weight",
+    "decoder.layers.{index}.linear1.bias",
+    "decoder.layers.{index}.linear2.weight",
+    "decoder.layers.{index}.linear2.bias",
+    "decoder.layers.{index}.norm1.weight",
+    "decoder.layers.{index}.norm1.bias",
+    "decoder.layers.{index}.norm2.weight",
+    "decoder.layers.{index}.norm2.bias",
+    "decoder.layers.{index}.norm3.weight",
+    "decoder.layers.{index}.norm3.bias",
+]
+ENCODER_PARAM_FQNS = [
+    "encoder.layers.{index}.self_attn.in_proj_weight",
+    "encoder.layers.{index}.self_attn.in_proj_bias",
+    "encoder.layers.{index}.self_attn.out_proj.weight",
+    "encoder.layers.{index}.self_attn.out_proj.bias",
+    "encoder.layers.{index}.linear1.weight",
+    "encoder.layers.{index}.linear1.bias",
+    "encoder.layers.{index}.linear2.weight",
+    "encoder.layers.{index}.linear2.bias",
+    "encoder.layers.{index}.norm1.weight",
+    "encoder.layers.{index}.norm1.bias",
+    "encoder.layers.{index}.norm2.weight",
+    "encoder.layers.{index}.norm2.bias",
+]
+TOTAL_NUM_PREFETCH = 12
+ENCODER_BEGIN_INDEX_FOR_PRE = 6
+ENCODER_BEGIN_INDEX_FOR_POST = 5
+ENCODER_PREFETCH_NUM = 5
+
+if not dist.is_available():
+    print("Distributed not available, skipping tests", file=sys.stderr)
+    sys.exit(0)
+
+if TEST_WITH_DEV_DBG_ASAN:
+    print(
+        "Skip dev-asan as torch + multiprocessing spawn have known issues",
+        file=sys.stderr,
+    )
+    sys.exit(0)
+
+
+@no_type_check
+def get_flat_param_fqns(state: _FSDPState, handle: FlatParamHandle) -> None:
+    if handle is None:
+        return None
+    param_to_fqn = state._exec_order_data.param_to_fqn
+    handle_params = handle.flat_param._params  # only populated for use_orig_params
+    param_fqns = [
+        param
+        for param_list in [param_to_fqn[p] for p in handle_params]
+        for param in param_list
+    ]
+    return param_fqns
+
+
+class TestBackwardPrefetch(FSDPTest):
+    @property
+    def world_size(self):
+        return 2
+
+    def _dist_train(self, backward_prefetch=BackwardPrefetch.BACKWARD_PRE):
+        rank = self.rank
+        orig_get_handle_to_prefetch = _get_handle_to_prefetch
+
+        torch.manual_seed(0)
+        policy = ModuleWrapPolicy(
+            {nn.TransformerEncoderLayer, nn.TransformerDecoderLayer}
+        )
+        model = FSDP(
+            nn.Transformer(d_model=1024, nhead=8, device="cuda"),
+            device_id=torch.cuda.current_device(),
+            auto_wrap_policy=policy,
+            use_orig_params=True,
+            backward_prefetch=backward_prefetch,
+        )
+        optim = torch.optim.SGD(model.parameters(), lr=1e-2)
+
+        # prepare input
+        torch.manual_seed(rank + 1)
+        src = torch.randn((10, 1, 1024), device="cuda")
+        tgt = torch.randn((20, 1, 1024), device="cuda")
+
+        # monkey patch
+        flat_param_fqns_array = []
```

Comment:
super nit: When I have a `List[List[str]]` I tend to prefix the name with `all_`. In this case, I would say `all_handle_fqns: List[List[str]] = []`, and then each element would be `handles_fqns: List[str]`.

Maybe `array` is not as popular terminology in Python unless using an explicit array data structure (e.g. numpy or the array library)?